Support retrieval of FV variable gradients with adCoupledGradient

lindsayad opened the pull request:
Simply populates the `MooseVariableDataFV::_ad_grad_u` member with the
result of a gradient reconstruction call to
`MooseVariableFV::adGradSln(elem)`.

Closes #16963

lindsayad committed:
Support retrieval of FV variable gradients with adCoupledGradient

Simply populates the `MooseVariableDataFV::_ad_grad_u` member with the
result of a gradient reconstruction call to
`MooseVariableFV::adGradSln(elem)`.

Closes #16963

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/681026/) on cea8907 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16964/site)

This comment will be updated on new commits.

lindsayad committed:
Build face information when calling Moose::faceInfo(elem, side) overload

Only build if we don't already have a clean build

lindsayad committed:
Update MOOSE-system data after the mesh has changed

If the mesh has changed, then, for instance, we need to rebuild our
FaceInfo data and the faceType data for reach variable for each FaceInfo

lindsayad committed:
Remove extraneous meshChanged calls

These should be extraneous because there appears to be no mesh operation
immediately before the calls *other* than possible uniform refinement
during restart...and within that uniform refine call, e.g.
`Adaptivity::uniformRefineWithProjection`, there is already a call to
`SubProblem::meshChanged` so we don't need to repeat ourselves

lindsayad committed:
Remove gradients from FVFluxKernel

These gradients are not used currently by any FVFluxKernel derived
object and I don't really anticipate that they ever would be. I can see
a user requesting the face gradient (for diffusion for example), but
that's not what we were supplying anyways. Requesting gradients that are
not used induces unnecessary computational expense and even worse can
result in segmentation faults if there is insufficient ghosting (as was
the case as soon as we actually started computing the gradients because
by default we only request one layer of ghosting for FV calculations but
we need two as soon as we start asking for a face gradient for a
neighbor cell centroid gradient. FVDiffusion does it right and asks for
two layers of ghosting)

rwcarlsen commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -3177,6 +3177,9 @@ MooseMesh::buildFaceInfo()
 const FaceInfo *
 MooseMesh::faceInfo(const Elem * elem, unsigned int side) const
 {
+  if (_face_info_dirty)
```
This check is redundant with the one in the `buildFaceInfo` func.

lindsayad replied:
Ok removed!
